### ISSUE TITLE
NTT Pipeline in FuTIL

### ISSUE THREAD

**Issue by cgyurgyik** [cucapra/futil#358]
## Algorithm
This is a FuTIL-generated pipeline for  the Cooley-Tukey algorithm that avoids the bit-reversal step in the on-the-fly calculation by using `phis` in bit reversed order. This algorithm (below) is presented in [Speeding up the Number Theoretic Transform for Faster Ideal Lattice-Based Cryptography](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/05/RLWE-1.pdf).

![image](https://user-images.githubusercontent.com/37983775/105499568-a9931700-5c8f-11eb-8183-31529a1bc323.png)

## Preconditions
- N = length of `a` is a power of 2.
- `q` meets the following criteria:
(1) q is a prime number.
(2) q >= max(`a`) and q >= N
(3) q is equivalent to 1 mod 2N.
- `phis` are powers of `phi` in bit-reversed order

## Hardware Used

2 * `std_mem_d1_ext` (`a`, `phis`)

4 * N `std_reg` for parallelization.
- `A` registers are used to save the output of calculations.
- `r` registers are used within the calculation.
- `mul` registers are used to save the output of the multiply with `phi`.
- `phi` registers are used to save the value of `phi` for each index.

N * `std_smod_pipe`
N * `std_smult_pipe` ***
N/2 * `std_sadd`
N/2 * `std_ssub`
*** We should be able to reduce this to `N / 2`.
 
Currently, the modulus value `q` is just hard-wired to be a constant, e.g. if `q = 97`, then I just do `mod_pipe.right = 32'd97`, rather than using N * `std_const` as well. Perhaps there's some negative implications to this that I'm missing?

## Control
There are several stages in the control flow on top of the pipeline stages one would expect in NTT. For N=4, it would look something like this:
```
control {
  seq {
    // Write from memory to registers.
    seq { preamble_0; preamble_1; preamble_2; preamble_3; }

    par {
      seq { stage0_row0_mul; stage0_row0_mod; }
      seq { stage0_row1_mul; stage0_row1_mod; }
      seq { stage0_row2_mul; stage0_row2_mod; }
      seq { stage0_row3_mul; stage0_row3_mod; }
    }
    // Update intermediary registers.
    par { precursor_0; precursor_1; precursor_2; precursor_3; }
    
    par {
      seq { stage1_row0_mul; stage1_row0_mod; }
      seq { stage1_row1_mul; stage1_row1_mod; }
      seq { stage1_row2_mul; stage1_row2_mod; }
      seq { stage1_row3_mul; stage1_row3_mod; }
    }
    // Write from registers back to memory.
    seq { epilogue_0; epilogue_1; epilogue_2; epilogue_3; }
  }
```
Let’s go through each of these steps.
### Preamble
The preamble essentially takes each value X in `a` and `phi` and writes it to a register `rX` and `phiX` respectively. This allows us to compute each stage in parallel. Before, we were blocked on the number of accesses to memory. This is done sequentially, so should take N cycles.

### Pipeline stage
This is the meat of the NTT algorithm. Here, we compute the necessary calculations for the butterfly unit. Since each calculation requires a multiply followed by a combinational `add` or `sub` with modulus, we split each stage into two pieces:
```
seq { 
  stage0_row0_mul;    // V = a[j] * phi[k]
  stage0_row0_mod;   // (U +- V) % q
}
```
The output of the `mul` group is saved in an intermediary register `mul` to be used in the `mod` group. Since all values of `a` and `phi` have been written to registers, we can do this in parallel. This is also the most expensive part of the pipeline. The output of the `mod` group writes to an intermediary register `A`.

*One thing to note is that we are using N `std_smult_pipe`s. I think this only needs to be N / 2, since the calculation `V` is used twice in each stage of the pipeline. This won’t make anything faster since it is in parallel anyways, but it’ll reduce the hardware count.

An example of  N=4 NTT pipeline stage calculations would be the following:
```
// +---+-----------------------+-----------------------+
// | a |        Stage 0        |        Stage 1        |
// +---+-----------------------+-----------------------+
// | 0 | a[0] + a[2] * phis[1] | a[0] + a[1] * phis[2] |
// | 1 | a[1] + a[3] * phis[1] | a[0] - a[1] * phis[2] |
// | 2 | a[0] - a[2] * phis[1] | a[2] + a[3] * phis[3] |
// | 3 | a[1] - a[3] * phis[1] | a[2] - a[3] * phis[3] |
// +---+-----------------------+-----------------------+
```

### Precursor
Before every stage except the first, there is a set of N precursor groups that run in parallel. They do one thing: Update the intermediary registers `r0…r(N-1)` with the new values produced from the last stage, which are saved in `A0…A(N-1)`.

### Epilogue
Finally, the epilogue is N groups that take the final outputs of the final stage (saved in, as usual, `A0…A(N-1)`) and puts them in the original array `a`. Since we are limited to one access at a time to the memory, this is done sequentially, and should take N cycles.

## FuTIL generator
I wrote a quick and dirty Python script to conduct this. Given parameters `N`, `input_bitwidth`, and `q`, it will output a FuTIL program, as well as a pretty-printed table that describes what calculates occur at what stage, as seen above.

## Example for `N`= 4, `q`= 97, and `input_bitwidth` = 32
```
// +---+-----------------------+-----------------------+
// | a |        Stage 0        |        Stage 1        |
// +---+-----------------------+-----------------------+
// | 0 | a[0] + a[2] * phis[1] | a[0] + a[1] * phis[2] |
// | 1 | a[1] + a[3] * phis[1] | a[0] - a[1] * phis[2] |
// | 2 | a[0] - a[2] * phis[1] | a[2] + a[3] * phis[3] |
// | 3 | a[1] - a[3] * phis[1] | a[2] - a[3] * phis[3] |
// +---+-----------------------+-----------------------+
import "primitives/std.lib";
component main() -> () {
  cells {
    a = prim std_mem_d1_ext(32, 4, 3);
    phis = prim std_mem_d1_ext(32, 4, 3);
    r0 = prim std_reg(32);
    r1 = prim std_reg(32);
    r2 = prim std_reg(32);
    r3 = prim std_reg(32);
    A0 = prim std_reg(32);
    A1 = prim std_reg(32);
    A2 = prim std_reg(32);
    A3 = prim std_reg(32);
    mul0 = prim std_reg(32);
    mul1 = prim std_reg(32);
    mul2 = prim std_reg(32);
    mul3 = prim std_reg(32);
    phi0 = prim std_reg(32);
    phi1 = prim std_reg(32);
    phi2 = prim std_reg(32);
    phi3 = prim std_reg(32);
    mod_pipe0 = prim std_smod_pipe(32);
    mod_pipe1 = prim std_smod_pipe(32);
    mod_pipe2 = prim std_smod_pipe(32);
    mod_pipe3 = prim std_smod_pipe(32);
    mult_pipe0 = prim std_smult_pipe(32);
    mult_pipe1 = prim std_smult_pipe(32);
    mult_pipe2 = prim std_smult_pipe(32);
    mult_pipe3 = prim std_smult_pipe(32);
    add0 = prim std_sadd(32);
    add1 = prim std_sadd(32);
    sub0 = prim std_ssub(32);
    sub1 = prim std_ssub(32);
  }
  wires {
    group preamble_0 {
      a.addr0 = 3'd0;
      phis.addr0 = 3'd0;
      r0.write_en = 1'd1;
      r0.in = a.read_data;
      phi0.write_en = 1'd1;
      phi0.in = phis.read_data;
      preamble_0[done] = r0.done & phi0.done? 1'd1;
    }
    group preamble_1 {
      a.addr0 = 3'd1;
      phis.addr0 = 3'd1;
      r1.write_en = 1'd1;
      r1.in = a.read_data;
      phi1.write_en = 1'd1;
      phi1.in = phis.read_data;
      preamble_1[done] = r1.done & phi1.done? 1'd1;
    }
    group preamble_2 {
      a.addr0 = 3'd2;
      phis.addr0 = 3'd2;
      r2.write_en = 1'd1;
      r2.in = a.read_data;
      phi2.write_en = 1'd1;
      phi2.in = phis.read_data;
      preamble_2[done] = r2.done & phi2.done? 1'd1;
    }
    group preamble_3 {
      a.addr0 = 3'd3;
      phis.addr0 = 3'd3;
      r3.write_en = 1'd1;
      r3.in = a.read_data;
      phi3.write_en = 1'd1;
      phi3.in = phis.read_data;
      preamble_3[done] = r3.done & phi3.done? 1'd1;
    }
    group precursor_0 {
      r0.in = A0.out;
      r0.write_en = 1'd1;
      precursor_0[done] = r0.done ? 1'd1;
    }
    group precursor_1 {
      r1.in = A1.out;
      r1.write_en = 1'd1;
      precursor_1[done] = r1.done ? 1'd1;
    }
    group precursor_2 {
      r2.in = A2.out;
      r2.write_en = 1'd1;
      precursor_2[done] = r2.done ? 1'd1;
    }
    group precursor_3 {
      r3.in = A3.out;
      r3.write_en = 1'd1;
      precursor_3[done] = r3.done ? 1'd1;
    }
    group s0_r0_mul {
      mult_pipe0.left = phi1.out;
      mult_pipe0.right = r2.out;
      mult_pipe0.go = !mult_pipe0.done ? 1'd1;
      mul0.write_en = mult_pipe0.done;
      mul0.in = mult_pipe0.out;
      s0_r0_mul[done] = mul0.done ? 1'd1;
    }
    group s0_r0_mod {
      add0.left = r0.out;
      add0.right = mul0.out;
      mod_pipe0.left = add0.out;
      mod_pipe0.right = 32'd97;
      mod_pipe0.go = !mod_pipe0.done ? 1'd1;
      A0.write_en = mod_pipe0.done;
      A0.in = mod_pipe0.out;
      s0_r0_mod[done] = A0.done ? 1'd1;
    }
    group s0_r1_mul {
      mult_pipe1.left = phi1.out;
      mult_pipe1.right = r3.out;
      mult_pipe1.go = !mult_pipe1.done ? 1'd1;
      mul1.write_en = mult_pipe1.done;
      mul1.in = mult_pipe1.out;
      s0_r1_mul[done] = mul1.done ? 1'd1;
    }
    group s0_r1_mod {
      add1.left = r1.out;
      add1.right = mul1.out;
      mod_pipe1.left = add1.out;
      mod_pipe1.right = 32'd97;
      mod_pipe1.go = !mod_pipe1.done ? 1'd1;
      A1.write_en = mod_pipe1.done;
      A1.in = mod_pipe1.out;
      s0_r1_mod[done] = A1.done ? 1'd1;
    }
    group s0_r2_mul {
      mult_pipe2.left = phi1.out;
      mult_pipe2.right = r2.out;
      mult_pipe2.go = !mult_pipe2.done ? 1'd1;
      mul2.write_en = mult_pipe2.done;
      mul2.in = mult_pipe2.out;
      s0_r2_mul[done] = mul2.done ? 1'd1;
    }
    group s0_r2_mod {
      sub0.left = r0.out;
      sub0.right = mul2.out;
      mod_pipe2.left = sub0.out;
      mod_pipe2.right = 32'd97;
      mod_pipe2.go = !mod_pipe2.done ? 1'd1;
      A2.write_en = mod_pipe2.done;
      A2.in = mod_pipe2.out;
      s0_r2_mod[done] = A2.done ? 1'd1;
    }
    group s0_r3_mul {
      mult_pipe3.left = phi1.out;
      mult_pipe3.right = r3.out;
      mult_pipe3.go = !mult_pipe3.done ? 1'd1;
      mul3.write_en = mult_pipe3.done;
      mul3.in = mult_pipe3.out;
      s0_r3_mul[done] = mul3.done ? 1'd1;
    }
    group s0_r3_mod {
      sub1.left = r1.out;
      sub1.right = mul3.out;
      mod_pipe3.left = sub1.out;
      mod_pipe3.right = 32'd97;
      mod_pipe3.go = !mod_pipe3.done ? 1'd1;
      A3.write_en = mod_pipe3.done;
      A3.in = mod_pipe3.out;
      s0_r3_mod[done] = A3.done ? 1'd1;
    }
    group s1_r0_mul {
      mult_pipe0.left = phi2.out;
      mult_pipe0.right = r1.out;
      mult_pipe0.go = !mult_pipe0.done ? 1'd1;
      mul0.write_en = mult_pipe0.done;
      mul0.in = mult_pipe0.out;
      s1_r0_mul[done] = mul0.done ? 1'd1;
    }
    group s1_r0_mod {
      add0.left = r0.out;
      add0.right = mul0.out;
      mod_pipe0.left = add0.out;
      mod_pipe0.right = 32'd97;
      mod_pipe0.go = !mod_pipe0.done ? 1'd1;
      A0.write_en = mod_pipe0.done;
      A0.in = mod_pipe0.out;
      s1_r0_mod[done] = A0.done ? 1'd1;
    }
    group s1_r1_mul {
      mult_pipe1.left = phi2.out;
      mult_pipe1.right = r1.out;
      mult_pipe1.go = !mult_pipe1.done ? 1'd1;
      mul1.write_en = mult_pipe1.done;
      mul1.in = mult_pipe1.out;
      s1_r1_mul[done] = mul1.done ? 1'd1;
    }
    group s1_r1_mod {
      sub0.left = r0.out;
      sub0.right = mul1.out;
      mod_pipe1.left = sub0.out;
      mod_pipe1.right = 32'd97;
      mod_pipe1.go = !mod_pipe1.done ? 1'd1;
      A1.write_en = mod_pipe1.done;
      A1.in = mod_pipe1.out;
      s1_r1_mod[done] = A1.done ? 1'd1;
    }
    group s1_r2_mul {
      mult_pipe2.left = phi3.out;
      mult_pipe2.right = r3.out;
      mult_pipe2.go = !mult_pipe2.done ? 1'd1;
      mul2.write_en = mult_pipe2.done;
      mul2.in = mult_pipe2.out;
      s1_r2_mul[done] = mul2.done ? 1'd1;
    }
    group s1_r2_mod {
      add1.left = r2.out;
      add1.right = mul2.out;
      mod_pipe2.left = add1.out;
      mod_pipe2.right = 32'd97;
      mod_pipe2.go = !mod_pipe2.done ? 1'd1;
      A2.write_en = mod_pipe2.done;
      A2.in = mod_pipe2.out;
      s1_r2_mod[done] = A2.done ? 1'd1;
    }
    group s1_r3_mul {
      mult_pipe3.left = phi3.out;
      mult_pipe3.right = r3.out;
      mult_pipe3.go = !mult_pipe3.done ? 1'd1;
      mul3.write_en = mult_pipe3.done;
      mul3.in = mult_pipe3.out;
      s1_r3_mul[done] = mul3.done ? 1'd1;
    }
    group s1_r3_mod {
      sub1.left = r2.out;
      sub1.right = mul3.out;
      mod_pipe3.left = sub1.out;
      mod_pipe3.right = 32'd97;
      mod_pipe3.go = !mod_pipe3.done ? 1'd1;
      A3.write_en = mod_pipe3.done;
      A3.in = mod_pipe3.out;
      s1_r3_mod[done] = A3.done ? 1'd1;
    }
    group epilogue_0 {
      a.addr0 = 3'd0;
      a.write_en = 1'd1;
      a.write_data = A0.out;
      epilogue_0[done] = a.done ? 1'd1;
    }
    group epilogue_1 {
      a.addr0 = 3'd1;
      a.write_en = 1'd1;
      a.write_data = A1.out;
      epilogue_1[done] = a.done ? 1'd1;
    }
    group epilogue_2 {
      a.addr0 = 3'd2;
      a.write_en = 1'd1;
      a.write_data = A2.out;
      epilogue_2[done] = a.done ? 1'd1;
    }
    group epilogue_3 {
      a.addr0 = 3'd3;
      a.write_en = 1'd1;
      a.write_data = A3.out;
      epilogue_3[done] = a.done ? 1'd1;
    }
  }
  control {
    seq {
      seq {
        preamble_0;
        preamble_1;
        preamble_2;
        preamble_3;
      }
      par {
        seq { s0_r0_mul; s0_r0_mod; }
        seq { s0_r1_mul; s0_r1_mod; }
        seq { s0_r2_mul; s0_r2_mod; }
        seq { s0_r3_mul; s0_r3_mod; }
      }
      par {
        precursor_0;
        precursor_1;
        precursor_2;
        precursor_3;
      }
      par {
        seq { s1_r0_mul; s1_r0_mod; }
        seq { s1_r1_mul; s1_r1_mod; }
        seq { s1_r2_mul; s1_r2_mod; }
        seq { s1_r3_mul; s1_r3_mod; }
      }
      seq {
        epilogue_0;
        epilogue_1;
        epilogue_2;
        epilogue_3;
      }
    }
  }
}
```

**Comment by cgyurgyik** [comment]
cc @rachitnigam since you've probably already answered and/or solved this.

So I’m doing something like this currently for a multiply followed by modulus:
```
group do_mul_op {
  ...
  mul_temp.in = mult_pipe.out; // Save product in a register.
}

group do_mod_op {
  ...
  mod_pipe.right = mul_temp.out; // Use product.
}

// Control
...
seq { do_mul_op; do_mod_op; }
```
However, this does require using a temporary register between the multiply and the modulus operations. 

Obviously it’d be great to avoid using a temporary register here. Just using `mult_pipe.out` in `do_mod_op` won’t work since the signal isn’t visible outside the group. I think this is exactly why the `invoke` operator was created. Don’t I want to do something like:
```
group do_mod_op {
  mod_pipe.right = mult0.out; // Use product; avoids a temporary register.
}

...

// Control
seq { 
  invoke mult0(...)();
  do_mod_op;
}
```

In fact, we could probably just do the same thing with the group `do_mod_op` as well, and feed the output of `mult0` to the arguments of `mod0`. However, I still need to save the output of `do_mod_op` to a register for use in the next stage, so a group after “both” invokes is still required.

**Comment by rachitnigam** [comment]
TLDR: Your approach is almost the best we can do in this case. The only possible optimization here to remove the register b/w mult and mod is sequence them within the group. 

You've run into two problems here:
1. The interface for `mult` and `mod` is not `invoke`-able (#331). I'm going to refer you to the issue for an explanation about this.
2. The compiler cannot reason about how long the signal from `mult` is alive for.

The hand optimization that you can do here—to eliminate the register b/w mult and mod would look like this:
```
group mult_mod_fused {
  // Run mult exactly once. mult.go = !mult.done is going to repeatedly run mult.
  mult.go = !once.out;
  once.out = mult.done;
  once.write_en = mult.done;
  // args to mult ...

  // Start mod on the cycle when the signal from mult is visible
  mod.go = mult.done || once.out;
  mod.left = mult.out; // this signal is valid on the first cycle. After that mod "saves" it.
  mod.right = ...
  mult_mod_fuse[done] = mod.done;
}
```

The compiler cannot reason about this optimization. It does not have the following pieces of information:
1. `mult.done` is alive for the same cycle as `done`. The compiler always has to assume that signals output signals are dead.
2. `mod.left` is *only* needs to be driven for one cycle. The `mod` component "saves" the signal after that cycle for use.

All of this goes back to the discussion we were having on Tuesday about the liveness of signals. The is *mostly* ignorant of which signals are alive when. The one thing it does know is that the `out` signal of registers is always alive. An extension to the Calyx compiler that reasons about the liveness of signals would allow these kinds of "chaining" optimizations. That's what we want to build next.

CC @sampsyo @sgpthomas @tedbauer since you might find this discussion interesting

**Comment by cgyurgyik** [comment]
Thanks for the long write-up.

> `mult.done` is alive for the same cycle as `done`. The compiler always has to assume that signals output signals are dead.

What is the second `done` you're referring to?

Ok, so I was assuming that `mult` and `mod` are` invoke`-able. #331 says that is not the case. it seems that problem 2, signal aliveness, is solved if we make it so?

One thing that I didn't mention here is that each stage only requires `N/2` multiplications, since there are duplicates.
Eventually, I only want to use `N / 2` `mult_pipe`s, which means that I'll want the signal connected to two different `mod_pipe`s. By guaranteeing the signal will last after invocation, and placing it before said `mod_pipe`s in a `seq`, this seems correct.

**Comment by rachitnigam** [comment]
Ah, typo.  I meant to say: `mult.out` is available when `mult.done` is high.

> Ok, so I was assuming that mult and mod are invoke-able. #331 says that is not the case. it seems that problem 2, signal aliveness, is solved if we make it so?

The "solution" to #331 is introducing a register into the `mult_pipe` primitive. Morally, this doesn't fix your problem of extra registers. It just gives you a new interface that lets you use values after `invoke`-ing a primitive. The solution that lets us implement the hand written optimization I showed you in the compiler requires more work.

**Comment by cgyurgyik** [comment]
A few more questions to digest this in its entirety.
1. In your example above, isn't `once` an extra  1-bit register as well?
2. You said:
> mod.left is only needs to be driven for one cycle. 

Curiously, is there anything wrong with "driving" it for longer? I'd assume no, since we write these to a register within the primitive.

Thanks for the responses!!

**Comment by rachitnigam** [comment]
> In your example above, isn't once an extra 1-bit register as well?

Yes

> Curiously, is there anything wrong with "driving" it for longer? I'd assume no, since we write these to a register within the primitive.

No. This is what happens currently. The question is what can you *assume* given a black-box component. Right now, all components using another component have to assume:
1. All input signals need to be driven as long as `go` is driven
2. No output signals are available after `done` signal is high

The `invoke` operator gives a very weak guarantee: it says that it's on the frontend to assume that the output ports it's reading from after the `invoke` are not dead.

**Comment by sampsyo** [comment]
Just wanted to say that this conversation I unfortunately read much later was super interesting! Indeed, this is a great motivator for more nuanced reasoning about what signals are available when as part of the "calling convention" for statically timed components…
